File: backend/src/data/CanvasAnnotation.java
package data;

import java.net.URI;

import javax.xml.namespace.QName;

public class CanvasAnnotation {

	private String namespaceURI;
	private String localPart;
	private String prefix;
	private String nestedNamespaceURI;
	private String nestedLocalPart;
	private String nestedPrefix;
	private URI nestedURI;
	private String stringValue;
	private URI uriValue;
	private CanvasAnnotation[] annotations;

	public void setQName(QName qname) {
		this.namespaceURI = qname.getNamespaceURI();
		this.localPart = qname.getLocalPart();
		this.prefix = qname.getPrefix();
	}
	
	public void setNestedQName(QName nestedQName) {
		this.nestedNamespaceURI = nestedQName.getNamespaceURI();
		this.nestedLocalPart = nestedQName.getLocalPart();
		this.nestedPrefix = nestedQName.getPrefix();
	}
	
	public String getNamespaceURI() {
		return namespaceURI;
	}

	public void setNamespaceURI(String namespaceURI) {
		this.namespaceURI = namespaceURI;
	}

	public String getLocalPart() {
		return localPart;
	}

	public void setLocalPart(String localPart) {
		this.localPart = localPart;
	}

	public String getPrefix() {
		return prefix;
	}

	public void setPrefix(String prefix) {
		this.prefix = prefix;
	}

	public String getNestedNamespaceURI() {
		return nestedNamespaceURI;
	}

	public void setNestedNamespaceURI(String nestedNamespaceURI) {
		this.nestedNamespaceURI = nestedNamespaceURI;
	}

	public String getNestedLocalPart() {
		return nestedLocalPart;
	}

	public void setNestedLocalPart(String nestedLocalPart) {
		this.nestedLocalPart = nestedLocalPart;
	}

	public String getNestedPrefix() {
		return nestedPrefix;
	}

	public void setNestedPrefix(String nestedPrefix) {
		this.nestedPrefix = nestedPrefix;
	}
	
	public URI getNestedURI() {
		return this.nestedURI;
	}
	
	public void setNestedURI(URI nestedURI) {
		this.nestedURI = nestedURI;
	}
	
	public String getStringValue() {
		return stringValue;
	}

	public void setStringValue(String stringValue) {
		this.stringValue = stringValue;
	}

	public URI getUriValue() {
		return uriValue;
	}

	public void setUriValue(URI uriValue) {
		this.uriValue = uriValue;
	}

	public CanvasAnnotation[] getAnnotations() {
		return annotations;
	}

	public void setAnnotations(CanvasAnnotation[] annotations) {
		this.annotations = annotations;
	}

}

File: frontend/src/app/canvasAnnotation.ts

export class CanvasAnnotation {
  namespaceURI: string;
  localPart: string;
  prefix: string;
  nestedNamespaceURI: string;
  nestedLocalPart: string;
  nestedPrefix: string;
  nestedURI: string;
  stringValue: string;
  uriValue: string;
  annotations: CanvasAnnotation[];

  encode(enc: any) {
    let node = enc.document.createElement('CanvasAnnotation');
    if (this.namespaceURI)
      node.setAttribute("namespaceURI", this.namespaceURI);
    if (this.localPart)
      node.setAttribute("localPart", this.localPart);
    if (this.prefix)
      node.setAttribute("prefix", this.prefix);
    if(this.nestedNamespaceURI)
      node.setAttribute("nestedNamespaceURI", this.nestedNamespaceURI);
    if(this.nestedLocalPart)
      node.setAttribute("nestedLocalPart", this.nestedLocalPart);
    if(this.nestedPrefix)
      node.setAttribute("nestedPrefix", this.nestedPrefix);
    if(this.nestedURI) 
      node.setAttribute("nestedURI", this.nestedURI);
    if (this.stringValue)
      node.setAttribute("stringValue", this.stringValue);
    if (this.uriValue)
      node.setAttribute("uriValue", this.uriValue);
    if (this.annotations) {
      let annotationsNode = enc.encode(this.annotations);
      annotationsNode.setAttribute("as", "annotations");
      node.appendChild(annotationsNode);
    }

    return node;
  }

}